### PR TITLE
Spark3: Support for Grouping Sets, `CUBE` and `ROLLUP` in `GROUP BY` clause of `SELECT` statement

### PR DESCRIPTION
**Pull request by R7L208**
### Brief summary of the change made

 - replace `GroupByClauseSegment` from ANSI to allow for `CUBE` and `ROLLUP`
 - created `WithCubeRollupClauseSegment` to handle the `WITH CUBE | WITH ROLLUP` syntax that occurs after a `GROUP BY`
 - Created `CubeRollupClauseSegment` based on Snowflake dialect to handle `CUBE | ROLLUP` inside of the `GROUP BY` clause
 - Created `GroupingSetsClauseSegment` based on Snowflake dialect to handle Grouping Sets within a `GROUP BY` clause
 - Created `GroupingExpressionList` for `CUBE`, `ROLLUP`, and Grouping Sets. Also based on existing class in Snowflake dialect.

### Are there any other side effects of this change that we should be aware of?
Based on a quick search:
- Cubes appear in Exasol, Postgres, Snowflake, and now Spark as implemented segments but are also listed in Hive, Ansi and BigQuery as Keywords.
- Rollups are implemented in Snowflake, Exasol, and Spark.
- Grouping Sets are also implemented in Snowflake, Exasol, and Spark.

It appears that segments (`CubeRollupClauseSegment`, `GroupingSetsClauseSegment`, `GroupingExpressionList`) from Spark, Snowflake, and Exasol are all copy/paste.

I think it makes sense to move these segments and needed grammar into Ansi dialect but would appreciate a second opinion before I do the lift & shift.

`WithCubeRollupClauseSegment` is unique to Spark3 for now, so that could be kept in Spark3 dialect.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by R7L208** [comment]
Seems that BigQuery only supports ROLLUP at the moment.
https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#group_by_clause

Hive Reference.
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+GroupBy#LanguageManualGroupBy-GroupingSets,Cubes,Rollups,andtheGROUPING__IDFunction

Exasol Reference (but need to scroll to down for cube/rollup/grouping sets)
https://docs.exasol.com/sql/select.htm

Snowflake docs.
https://docs.snowflake.com/en/sql-reference/constructs/group-by-cube.html
https://docs.snowflake.com/en/sql-reference/constructs/group-by-rollup.html
https://docs.snowflake.com/en/sql-reference/constructs/group-by-grouping-sets.html

File: src/sqlfluff/dialects/dialect_spark3.py [hunk 1]
```diff
@@ -1086,6 +1085,143 @@ class SelectStatementSegment(BaseSegment):
...
+    """
+
+    type = "with_cube_rollup_clause"
+
+    match_grammar = StartsWith(
+        Sequence(
+            "WITH",
+            OneOf("CUBE", "ROLLUP"),
+        ),
+        terminator=OneOf(
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_spark3.py:1141]: Why do you need this? It’s a simple enough statement so shouldn’t be confused with the next statement (which is the usual reason for adding a terminator clause).

So can this `match_clause` and `parse_clause`s be merged into one simple `match_clause`?

**Review by tunetheweb — COMMENTED** [review]
Good set of test case!

I do have one question though.

File: src/sqlfluff/dialects/dialect_spark3.py (hunk 1 above)
Comment (R7L208) [reply src/sqlfluff/dialects/dialect_spark3.py:1141]: yes, I'll make that update. They can be merged

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**